packages/eslint-plugin/tests/rules/require-types-exports.test.ts

```diff
@@ -0,0 +1,1217 @@
+    },
+
+    {
+      code: `
+        type Arg1 = number;
+        type Arg2 = string;
+
+        export function f({ a, b }: { a: Arg1; b: Arg2; c: number }): void {}
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 5,
+          column: 42,
+          endColumn: 46,
+          data: {
+            name: 'Arg1',
+          },
+        },
+        {
+          messageId: 'requireTypeExport',
+          line: 5,
+          column: 51,
+          endColumn: 55,
+          data: {
+            name: 'Arg2',
+          },
+        },
+      ],
+    },
+
+    {
+      code: `
+        type Arg1 = number;
+        type Arg2 = string;
+
+        export const f = ({ a, b }: { a: Arg1; b: Arg2; c: number }): void => {};
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 5,
+          column: 42,
+          endColumn: 46,
+          data: {
+            name: 'Arg1',
+          },
+        },
+        {
+          messageId: 'requireTypeExport',
+          line: 5,
+          column: 51,
+          endColumn: 55,
+          data: {
+            name: 'Arg2',
+          },
+        },
+      ],
+    },
+
+    {
+      code: `
+        type Arg = number;
+
+        export function f(...args: Arg[]): void {}
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 4,
+          column: 36,
+          endColumn: 39,
+          data: {
+            name: 'Arg',
+          },
+        },
+      ],
+    },
+
+    {
+      code: `
+        type Arg = number;
+
+        export const f = (...args: Arg[]): void => {};
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 4,
+          column: 36,
+          endColumn: 39,
+          data: {
+            name: 'Arg',
+          },
+        },
+      ],
+    },
+
+    {
+      code: `
+        type Arg = number;
+
+        export function f(a: Arg = 1): void {}
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 4,
+          column: 30,
+          endColumn: 33,
+          data: {
+            name: 'Arg',
+          },
+        },
+      ],
+    },
+
+    {
+      code: `
+        type Arg = number;
+
+        export const f = (a: Arg = 1): void => {};
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 4,
+          column: 30,
+          endColumn: 33,
+          data: {
+            name: 'Arg',
+          },
+        },
+      ],
+    },
+
+    // TODO: Find a resaonable way to handle this case
+    {
+      code: `
+        type Arg = number;
+
+        export function f<T extends Arg>(a: T): void {}
+      `,
+      errors: [
+        {
+          messageId: 'requireTypeExport',
+          line: 4,
+          column: 37,
+          endColumn: 39,
+          data: {
+            name: 'Arg',
+          },
+        },
+      ],
+    },
```

Comment:
Help would be greatly appreciated here ❤️

Reply:
Sorry, I'm not clear: what's the request?

Reply:
Have you tried using the scope manager? That's our normal go-to for answering questions like _"where is this referenced?"_.

https://typescript-eslint.io/play#ts=5.3.3&showAST=scope&fileType=.tsx&code=C4TwDgpgBAggTgcygXigOwK4FsBGE4DcAUERAB5gD2cwUAZhmgMbACWla9APACpTnAIaACYBnWIgB8ACgCGALig8AlIoBulVsKgBvKAF8iQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Reply:
Scope manager isn't for variables?
I used it here for something else:

https://github.com/StyleShit/typescript-eslint/blob/b309b51e1ae3246441e31e77724a5046f850f2d7/packages/eslint-plugin/src/rules/require-types-exports.ts#L82-L105

(or we're talking about different things? can you direct me to the APIs you're referring to? or to a code example in the codebase?)

Reply:
Yup exactly - it sees all (type and variables) and knows all (of their referenced scopes).

Reply:
(It's ok, you've already helped me a lot here with those few words 🙏)